### PR TITLE
[BIOMAGE-1677] Make aggregated pseudobulk matrix

### PR DESCRIPTION
**Pull request by gerbeldo**
# Description
Implements a function to aggregate count matrices, to use in pseudobulk differential expression.

# Details
#### URL to issue
https://biomage.atlassian.net/browse/BIOMAGE-1677

#### Link to staging deployment URL (or set N/A)
N/A

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [ ] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [x] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [x] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [x] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [x] Started end-to-end tests on the latest commit.

### Documentation updates
- [x] Relevant Github READMEs updated **or** no GitHub README updates required.
- [x] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [X] Photo of a cute animal attached to this PR.
Screen Shot 2022-02-01 at 14 22 07

### CODE REVIEW

File: r/R/make_pseudobulk_matrix.R [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+#' @return a SeuratObject with counts aggregated by sample for one cluster
+#' @export
+#'
+makePseudobulkMatrix <- function(scdata) {
+  counts <- scdata[["RNA"]]@counts
+  gene_annotations <- scdata@misc$gene_annotations
+
+  # create groups for aggregation
+  pbulk_groups <-
+    factor(paste(scdata$samples, scdata$custom, sep = "_"))
```
Comment (gerbeldo) [inline r/R/make_pseudobulk_matrix.R]: This should use a better separator. samples usually contain underscores.

File: r/R/make_pseudobulk_matrix.R [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  agg <- presto::sumGroups(counts, pbulk_groups, MARGIN = 1)
+  agg <- Matrix::Matrix(agg, sparse = TRUE)
+  agg <- Matrix::t(agg)
+
+  rownames(agg) <- rownames(counts)
+  colnames(agg) <- levels(pbulk_groups)
+
+  # recover metadata
+  pbulk_metadata <- colnames(agg)
+  pbulk_metadata <- data.table::tstrsplit(pbulk_metadata, split = "_")
```
Comment (gerbeldo) [inline r/R/make_pseudobulk_matrix.R]: same here when splitting

File: r/R/make_pseudobulk_matrix.R (hunk 1 above)
Comment (alexvpickering) [reply r/R/make_pseudobulk_matrix.R]: once you subset scdata you can just do the aggregation on samples

Comment (gerbeldo) [reply r/R/make_pseudobulk_matrix.R]: yeah! that was one of the things I wasn't sure about. that makes it simpler. I'll change it

File: r/R/make_pseudobulk_matrix.R [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+#' Also, a `samples` meta.data slot is required, specifying samples.
+#'
+#' @param scdata a SeuratObject,
+#'
+#' @return a SeuratObject with counts aggregated by sample for one cluster
+#' @export
+#'
+makePseudobulkMatrix <- function(scdata) {
+  # filter out cells not in base/background groups
+  scdata <- scdata[, !is.na(scdata@meta.data$custom)]
```
Comment (alexvpickering) [inline r/R/make_pseudobulk_matrix.R]:
```suggestion
  scdata <- scdata[, !is.na(scdata$custom)]
```

File: r/R/make_pseudobulk_matrix.R [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+  counts <- scdata[["RNA"]]@counts
+  gene_annotations <- scdata@misc$gene_annotations
+
+  groups <- factor(scdata@meta.data$samples)
```
Comment (alexvpickering) [inline r/R/make_pseudobulk_matrix.R]:
```suggestion
  groups <- factor(scdata$samples)
```

File: r/R/make_pseudobulk_matrix.R [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+#' @return a data.frame with a row per sample/custom
+#' @export
+#'
+#' @examples
+getSampleGroupKey <- function(scdata) {
+  metadata <- unique(dplyr::select(scdata@meta.data, samples, custom))
+  rownames(metadata) <- NULL
+
+  return(metadata)
+}
```
Comment (alexvpickering) [inline r/R/make_pseudobulk_matrix.R]:
```suggestion
  # aggregate over samples
  samples <- scdata$samples
  samples <- factor(samples, levels = unique(samples))

  agg <- presto::sumGroups(counts, samples, MARGIN = 1)
  agg <- Matrix::Matrix(agg, sparse = TRUE)
  agg <- Matrix::t(agg)

  #row/colnames are lost in aggregation
  rownames(agg) <- rownames(counts)
  colnames(agg) <- levels(samples)

  # recover original metadata
  metadata <- data.frame(samples = colnames(agg),
                         custom = scdata$custom[!duplicated(samples)],
                         row.names = colnames(agg))

  # create seurat, and add metadata
  pbulk <- Seurat::CreateSeuratObject(agg, meta.data = metadata)
  pbulk@misc$gene_annotations <- gene_annotations

  return(pbulk)
```

A couple of comments:
- have to be careful with letting levels be set by `factor` as they are by default sorted into increasing order. I think this is why you had the inner join and key lookup by you can probably avoid by just explicitly setting the levels to the order that items appear in
- other than that looks good, I just cleaned it up a bit

File: r/R/make_pseudobulk_matrix.R (hunk 4 above)
Comment (gerbeldo) [reply r/R/make_pseudobulk_matrix.R:19]: yep, explicit is better. It was easier to type when playing around, then it stuck!

File: r/tests/testthat/test-make-pseudobulk-matrix.R [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      sample == "s1" ~ "base",
+      sample == "s2" ~ "base",
+      sample == "s3" ~ "background",
+      sample == "s4" ~ "background"
+    )
+
+  # add cells not belonging to groups to compare
+  scdata$custom[floor(ncol(scdata) / 10) * 1:10] <- NA
+
+  scdata
```
Comment (alexvpickering) [inline r/tests/testthat/test-make-pseudobulk-matrix.R]:
```suggestion
  return(scdata)
```

File: r/tests/testthat/test-make-pseudobulk-matrix.R [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+
+test_that("makePseudobulkMatrix returns object of correct type and dims", {
+  scdata <- mock_scdata()
+  res <- makePseudobulkMatrix(scdata)
+
+  expected_rows <- nrow(scdata)
+  expected_cols <-
+    length(unique(scdata@meta.data$samples[!is.na(scdata@meta.data$custom)]))
```
Comment (alexvpickering) [inline r/tests/testthat/test-make-pseudobulk-matrix.R]:
```suggestion
    length(unique(scdata$samples[!is.na(scdata$custom)]))
```

File: r/tests/testthat/test-make-pseudobulk-matrix.R [hunk 8]
```diff
@@ -0,0 +1,86 @@
+mock_scdata <- function() {
+  scdata <- pbmc_small
```
Comment (alexvpickering) [inline r/tests/testthat/test-make-pseudobulk-matrix.R:3]:
```suggestion
  data("pbmc_small", package = "SeuratObject", envir = environment())
  scdata <- pbmc_small
```

**Review by alexvpickering — APPROVED** [review]
LGTM. see suggestions, nice work!

**Review by ogibson — APPROVED** [review]

**Pull request merged**